build-common/src/org/jetbrains/kotlin/incremental/LookupStorage.kt

```diff
@@ -302,31 +304,39 @@ private class TrackedLookupMap(private val lookupMap: LookupMap, private val tra
     operator fun get(key: LookupSymbolKey): Collection<Int>? = lookupMap[key]
 
     operator fun set(key: LookupSymbolKey, fileIds: Set<Int>) {
+        recordSet(key)
         lookupMap[key] = fileIds
-        recordAddedOrUpdatedKey(key)
     }
 
     fun append(key: LookupSymbolKey, fileIds: Collection<Int>) {
+        recordSet(key)
         lookupMap.append(key, fileIds)
-        recordAddedOrUpdatedKey(key)
     }
 
     fun remove(key: LookupSymbolKey) {
+        recordRemove(key)
         lookupMap.remove(key)
-        recordRemovedKey(key)
     }
 
-    private fun recordAddedOrUpdatedKey(key: LookupSymbolKey) {
-        if (trackChanges) {
-            removedKeys!!.remove(key)
-            addedOrUpdatedKeys!!.add(key)
+    private fun recordSet(key: LookupSymbolKey) {
+        if (!trackChanges) return
+        if (lookupMap[key] == null) {
+            if (key in removedKeys!!) {
+                removedKeys.remove(key)
```

Comment:
simplify with 
  val removed = removedKeys.remove(key)
  if (!removed) addedKeys.add(key)

Reply:
Hm that would be a bit faster but I find it slightly difficult to read.

```
            if (key in removedKeys!!) {
                removedKeys.remove(key) // Corner case
            } else {
                addedKeys!!.add(key) // Common case
            }
```            

With the above code, it's clear that we are treating a corner case vs. a common case differently.

With `val removed = removedKeys.remove(key)`, it's not clear why we immediately start removing the key. It takes a while to process that code and understand it. (Also, the saved condition check is only in the corner case, which rarely happens; in the common case they're the same.)